update cache directly after sources.list was changed

davidak opened the pull request:
fixes #7 

With this change we are able to change the `apt_ubuntu_mirror` without errors.

tested on Ubuntu 12.04, 14.04 and 16.04!

```
-deb mirror://mirrors.ubuntu.com/mirrors.txt precise multiverse
-deb-src mirror://mirrors.ubuntu.com/mirrors.txt precise multiverse
-deb mirror://mirrors.ubuntu.com/mirrors.txt precise-updates multiverse
-deb-src mirror://mirrors.ubuntu.com/mirrors.txt precise-updates multiverse
+deb http://de.archive.ubuntu.com/ubuntu/ precise multiverse
+deb-src http://de.archive.ubuntu.com/ubuntu/ precise multiverse
+deb http://de.archive.ubuntu.com/ubuntu/ precise-updates multiverse
+deb-src http://de.archive.ubuntu.com/ubuntu/ precise-updates multiverse

 # # N.B. software from this repository may not have been tested as
 # # extensively as that contained in the main release, although it includes


RUNNING HANDLER [tersmitten.apt : update] **************************************
ok: [dev-ansible-xenial] => {"cache_update_time": 1473243056, "cache_updated": true, "changed": false}
ok: [dev-ansible-trusty] => {"cache_update_time": 1473243057, "cache_updated": true, "changed": false}
ok: [dev-ansible-precise] => {"cache_update_time": 1473243057, "cache_updated": true, "changed": false}
```

noplanman commented:
Just stumbled in here and saw this.

If you're on Ansible 1.9+ you can also make use of the ternary filter like so:
```
- name: update configuration file - /etc/apt/sources.list
  template:
    src: "etc/apt/sources.list.{{ ansible_distribution }}.j2"
    dest: /etc/apt/sources.list
    owner: root
    group: root
    mode: 0644
  when: apt_manage_sources_list
  register: apt_config_updated        # be sure to register if this has changed.
  tags:
    - configuration
    - apt
    - apt-configuration

- name: update
  apt:
    update_cache: true
    # if the sources have been updated, set the cache_valid_time to 0, else use the set value.
    cache_valid_time: "{{ (apt_config_updated.changed) | ternary(0, apt_update_cache_valid_time) }}"
  when: apt_update
  tags:
    - configuration
    - apt
    - apt-update
```

Basically, if the sources have been updated, the apt cache is updated regardless of the set `cache_valid_time` (as it just sets it to 0).

Thought I'd just share that, as it's a smaller and more concise alternative.

Hope I understood the PR correctly, otherwise just take away from this message, that the ternary filter exists 😜

davidak commented:
@noplanman that looks elegant, but your role want's to be compatible with Ansible >= 1.6.

Using a handler is the obvious way to do it, like you learm from the Ansible docs.

noplanman commented:
@davidak Yes I know, unfortunately this nifty feature was only added in 1.9.

Probably not applicable for this role, as it seems to be very backwards compatible 👍

tersmitten commented on tasks/main.yml:
```diff
@@ -8,11 +8,14 @@
     - apt
     - apt-configuration
 
+- meta: flush_handlers
```
I rather have a task here, since this flushes all (open) handlers.

davidak replied:
yes, you are right.

i tried to add `apt_sources_list_file.changed` to your update task but nothing happens since the `apt_update_cache_valid_time` is in place.

any idea how i can reuse that task?

tersmitten commented:
What about changing the suggested ternary solution to `cache_valid_time: "{{ 0 if apt_config_updated.changed else apt_update_cache_valid_time }}"`. Can you guy test that?

davidak commented:
I updated and tested it.

What do you think about `when: ... apt_config_updated.changed`?

So it updates every time. Maybe someone wants to stop updating at all so sets `apt_update: false`?

tersmitten commented on tasks/main.yml:
```diff
@@ -16,8 +17,8 @@
-    cache_valid_time: "{{ apt_update_cache_valid_time }}"
-  when: apt_update
+    cache_valid_time: "{{ 0 if apt_config_updated is defined and apt_config_updated.changed else apt_update_cache_valid_time }}"
+  when: apt_update or apt_config_updated.changed
```
This will fail when `apt_config_updated` is undefined (in case `apt_manage_sources_list` is false)

davidak replied:
OK, i removed it.

Still works with `apt_update: true`.
